File: android/ClipSyncAndroid/app/src/main/java/com/clipsync/android/crypto/DeviceKeypair.kt
package com.clipsync.android.crypto

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import java.security.SecureRandom
import java.util.Base64

/**
 * This device's long-term X25519 identity keypair — the Android counterpart of
 * the Mac `DeviceKeypair`. Uses BouncyCastle so the raw 32-byte key format and
 * the X25519 agreement match CryptoKit byte-for-byte (see the crypto contract in
 * docs/superpowers/specs/2026-07-17-phase4a-mac-crypto-design.md).
 */
class DeviceKeypair private constructor(private val privateKey: X25519PrivateKeyParameters) {

    companion object {
        fun generate(): DeviceKeypair =
            DeviceKeypair(X25519PrivateKeyParameters(SecureRandom()))

        fun fromRawPrivateKey(raw: ByteArray): DeviceKeypair {
            require(raw.size == 32) { "X25519 private key must be 32 bytes" }
            return DeviceKeypair(X25519PrivateKeyParameters(raw, 0))
        }
    }

    /** 32-byte raw private key — persist this (Keystore); never send it. */
    val privateKeyRaw: ByteArray get() = privateKey.encoded

    /** 32-byte raw public key — safe to share; travels in `hello`. */
    val publicKeyRaw: ByteArray get() = privateKey.generatePublicKey().encoded

    /** Base64 of the public key, the exact form carried on the wire. */
    val publicKeyBase64: String get() = Base64.getEncoder().encodeToString(publicKeyRaw)

    /** X25519 agreement with a peer's raw 32-byte public key. */
    fun sharedSecret(peerPublicKey: ByteArray): ByteArray {
        require(peerPublicKey.size == 32) { "X25519 public key must be 32 bytes" }
        val agreement = X25519Agreement()
        agreement.init(privateKey)
        val secret = ByteArray(agreement.agreementSize)
        agreement.calculateAgreement(X25519PublicKeyParameters(peerPublicKey, 0), secret, 0)
        return secret
    }
}

File: mac/ClipSyncCore/Sources/ClipSyncCore/DeviceKeypair.swift
import Foundation
import CryptoKit

/// This device's long-term **X25519** identity keypair (see `docs/pairing-flow.md`).
///
/// Wraps CryptoKit's `Curve25519.KeyAgreement` keys and exposes the raw 32-byte
/// representations that travel in `hello` (public key) and get persisted in the
/// Keychain (private key). The keypair is symmetric per device, which is what
/// lets sync become bidirectional later without a protocol change.
public struct DeviceKeypair {
    public let privateKey: Curve25519.KeyAgreement.PrivateKey

    /// Generates a fresh keypair (first run).
    public init() {
        self.privateKey = Curve25519.KeyAgreement.PrivateKey()
    }

    /// Rehydrates a keypair from a stored 32-byte raw private key.
    public init(rawPrivateKey: Data) throws {
        self.privateKey = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: rawPrivateKey)
    }

    /// 32-byte raw private key — persist this (Keychain); never send it.
    public var privateKeyRaw: Data { privateKey.rawRepresentation }

    /// 32-byte raw public key — safe to share; travels in `hello`.
    public var publicKeyRaw: Data { privateKey.publicKey.rawRepresentation }

    /// The public key in the exact base64 form carried on the wire.
    public var publicKeyBase64: String { publicKeyRaw.base64EncodedString() }

    /// X25519 key agreement with a peer's raw 32-byte public key.
    public func sharedSecret(withPeerPublicKey peerPublicKey: Data) throws -> SharedSecret {
        let peer = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerPublicKey)
        return try privateKey.sharedSecretFromKeyAgreement(with: peer)
    }
}
